qualcommax: ipq807x: Xiaomi AX9000 add EMC2305 fan controller support

everything411 opened the pull request:
The Xiaomi AX9000 router has an integrated 6cm cooling fan controlled by an EMC2305 fan controller. This commit adds support for it using the in-tree emc2305 driver.

The driver will create a hwmon interface at `/sys/class/hwmon/hwmonN`. On Xiaomi AX9000, the EMC2305 typically appears as hwmon12.

Fan speed can be adjusted by writing to the pwm1 control file, where <fan_pwm_speed> is an integer between 0-255.
```
echo <fan_pwm_speed> > /sys/class/hwmon/hwmonN/pwm1
```

Thanks for @devrus91's work in https://forum.openwrt.org/t/openwrt-support-for-xiaomi-ax9000/98908/1935

robimarko commented:
Yeah, i2c-tools should not be included by default for sure, kmod is fine

everything411 commented:
Thanks! This is a test draft with i2c-tools for testing. I will clean up these two commits and submit the final version after testing is completed.

everything411 commented:
With https://github.com/openwrt/openwrt/pull/19137, emc2305 appears as /sys/class/hwmon/hwmon12; without it, emc2305 appears as /sys/class/hwmon/hwmon0.

aiamadeus commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-ax9000.dts:
```diff
@@ -121,6 +122,19 @@
+	emc2305@2f {
+		compatible = "microchip,emc2305";
+		reg = <0x2f>;
+		model_name = "emc2305";
```
Why do we need model_name?

aiamadeus commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-ax9000.dts:
```diff
@@ -121,6 +122,19 @@
+		reg = <0x2f>;
+		model_name = "emc2305";
+		#address-cells = <0x01>;
+		#size-cells = <0x00>;
```
```
#address-cells = <0x1>;
#size-cells = <0x0>;
```
Also keep a blank line.

aiamadeus commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-ax9000.dts:
```diff
@@ -121,6 +122,19 @@
+		emc2305_fan@0 {
+			reg = <0>;
+			name = "emc2305_fan";
+			pwm-output-mode = <1>;
```
These properties are useless, please refer to this:
https://github.com/torvalds/linux/blob/master/Documentation/devicetree/bindings/hwmon/microchip%2Cemc2305.yaml

everything411 commented:
Sorry for the late reply. I’ve been dealing with some personal matters lately and currently don’t have access to my AX9000, so I’m unable to conduct further testing for now.



> > With #19137, emc2305 appears as /sys/class/hwmon/hwmon12; without it, emc2305 appears as /sys/class/hwmon/hwmon0.
> 
> So what?
> 
> hwmon devices dont guarantee an ordering

The current PR code requires users to manually locate the hwmon number of the fan controller and set the fan speed.
Due to the specific hardware and .ko loading sequence on the AX9000, the hwmon interface for the EMC2305 fan controller always appears after the CPU thermal_zone is loaded but before the ath10k and ath11k drivers are loaded. This results in a relatively predictable behavior on the AX9000. 

For @aiamadeus's review, most of the dts is copied from @devrus91's work in the forum. I'm checking the documents and improve the format and useless part.

Regarding automatic speed control, I'm not entirely clear about how to properly integrate the dts for `pwm-fan` and `emc2305`. I’d like to seek @aiamadeus’s help on this matter.

robimarko commented:
> The current PR code requires users to manually locate the hwmon number of the fan controller and set the fan speed. Due to the specific hardware and .ko loading sequence on the AX9000, the hwmon interface for the EMC2305 fan controller always appears after the CPU thermal_zone is loaded but before the ath10k and ath11k drivers are loaded. This results in a relatively predictable behavior on the AX9000.

Again, the hwmon number is completely random, its just whatever was free when the driver was probed, you cannot rely on it.

> For @aiamadeus's review, most of the dts is copied from @devrus91's work in the forum. I'm checking the documents and improve the format and useless part.

DTS seems to rely on an out-of-tree driver and custom properties which dont exist upstream and are not parsed, as currently is, its not suitable for being merged

everything411 committed:
qualcommax: ipq807x: add EMC2305 fan controller support for Xiaomi AX9000

The Xiaomi AX9000 router has an integrated 6cm cooling fan controlled
by an EMC2305 fan controller. This commit adds support for it using the
in-tree emc2305 driver.

The driver will create a hwmon interface at `/sys/class/hwmon/hwmonN`.
Fan speed can be adjusted by writing to the pwm1 control file:
  echo <fan_pwm_speed> > /sys/class/hwmon/hwmonN/pwm1
  where <fan_pwm_speed> is an integer between 0-255

The Fan speed can be monitored via `sensors` command:
  emc2305-i2c-1-2f
  Adapter: QUP I2C adapter
  fan1:    1324 RPM
  pwm1:    26%

Forum link: https://forum.openwrt.org/t/openwrt-support-for-xiaomi-ax9000/98908/1935

Co-authored-by: Ruslan <ruslan.fedoseenko@protonmail.com>
Signed-off-by: Zheng Zhang <everything411@qq.com>

everything411 commented:
The `emc2305` driver is indeed in-tree and I've cleaned up the dts just now and it should contain not parsed part anymore.

Once  the the fan is bound to thermal-zones for automatic speed control, the manual access to /sys/class/hwmon/hwmonN is no longer needed. However I'm not familiar with thermal-zones and pwm-fans DTS, I'm currently not able to complete it. Therefore I will convert this PR to draft until someone helps me with that.

Thanks again for your review!

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-ax9000.dts:
```diff
@@ -121,6 +122,17 @@
+		#size-cells = <0x00>;
+
+		fan@0 {
+			reg = <0>;
```
Is this even being parsed by the kernel at all?

everything411 commented:
close since it is replaced by https://github.com/openwrt/openwrt/pull/22942
